app/assets/javascripts/admin/order_cycles/controllers/incoming_controller.js.coffee

```diff
@@ -9,19 +9,23 @@ angular.module('admin.orderCycles').controller 'AdminOrderCycleIncomingCtrl', ($
     enterprise = $scope.enterprises[exchange.enterprise_id]
     return enterprise.numVariants if enterprise.numVariants?
 
-    $scope.loadExchangeProducts(exchange)
-    return unless enterprise.supplied_products?
+    enterprise.numVariants = 0
+    params = { exchange_id: exchange.id, enterprise_id: exchange.enterprise_id, order_cycle_id: $scope.order_cycle.id, incoming: true}
+    ExchangeProduct.countVariants params, (variants_count) ->
+      enterprise.numVariants = variants_count
+      $scope.setSelectAllVariantsCheckboxValue(exchange, enterprise.numVariants)
 
-    enterprise.numVariants = $scope.countVariants(enterprise.supplied_products)
-
-  $scope.countVariants = (products) ->
-    return 0 unless products
-
-    numVariants = 0
-    for product in products
-      numVariants += product.variants.length
-    numVariants
+    return enterprise.numVariants
 
   $scope.addSupplier = ($event) ->
     $event.preventDefault()
     OrderCycle.addSupplier $scope.new_supplier_id
+
+  # To select all variants we first need to load them all from the server
```

Comment:
Is this the best way to handle selecting all variants? Can't we decouple the saving of the "selected" state from loading all the products?

So instead of fetching and loading all the available products in the exchange first, then sending another request to include all those products in the exchange (and save it), can't we just have an endpoint on the server-side that handles the task of "select all the available products in this exchange" and call that directly without doing the first step?

Reply:
yes! that's exactly #4487

Reply:
Ah sorry, I wasn't looking at all the issues at once.

Reply:
no problem, it's good you described the issue in your comment without looking at it :-)